{"metadata": {"language_info": {"name": "python", "version": "3.9.16", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}, "kernelspec": {"name": "python3", "display_name": "Python 3 (ipykernel)", "language": "python"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "import dxpy\nimport dxdata\nimport subprocess\nimport ast\nimport numpy as np\nimport pandas as pd\nimport glob \nimport os", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# Required dependencies \npip install pybgen\npip install zstandard", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "#split string input of multiple SNPs\ndef split_input(string_input):\n    if ':' in string_input:\n        parts = string_input.split(':')\n    elif '-' in string_input:\n        parts = string_input.split('-')\n    chrom = parts[0]\n    pos = int(parts[1])\n    ref = parts[2]\n    alt = parts[3]\n    if len(parts) > 4:\n        rsnum = parts[4]\n        return chrom,pos,ref,alt,rsnum\n    else:\n        return chrom, pos, ref, alt\n", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# Get Whole Exome Sequencing Dosages (per chromosome)\ndef getWESPath(chrom):\n    wes_path = '/mnt/project/Bulk/Exome sequences/Population level exome OQFE variants, BGEN format - final release/ukb23159_c' +chrom[3:]+ '_b0_v1.bgen'\n    return wes_path", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# Get TOPMED imputed dosages (per chromosome)\ndef getImputedPath(chrom):\n    imputed_path = '/mnt/project/Bulk/Imputation/Imputation from genotype (TOPmed)/ukb21007_c' + chrom[3:] + '_b0_v1.bgen'\n    return imputed_path", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# Return Sample Dataframe for Imputed dosages\ndef getSampleDF(chrom):\n    sample_path = '/mnt/project/Bulk/Imputation/Imputation from genotype (TOPmed)/ukb21007_c' + chrom[3:] + '_b0_v1.sample'\n    sample_df = pd.read_csv(sample_path, sep=' ')\n    sample_df = sample_df.drop(index=0)\n    sample_df = sample_df.drop(['missing', 'sex'], axis=1)\n    sample_df = sample_df.reset_index(drop=True)\n    sample_df = sample_df.rename(columns={\"ID_1\": \"FID\", \"ID_2\": \"IID\"}, errors=\"raise\")\n    return sample_df", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# Return Sample Dataframe for whole exome sequencing dosages\ndef getWESSampleDF(chrom):\n    sample_path = '/mnt/project/Bulk/Exome sequences/Population level exome OQFE variants, BGEN format - final release/ukb23159_c' + chrom[3:] + '_b0_v1.sample'\n    sample_df = pd.read_csv(sample_path, sep=' ')\n    sample_df = sample_df.drop(index=0)\n    sample_df = sample_df.drop(['missing', 'sex'], axis=1)\n    sample_df = sample_df.reset_index(drop=True)\n    sample_df = sample_df.rename(columns={\"ID_1\": \"FID\", \"ID_2\": \"IID\"}, errors=\"raise\")\n    return sample_df", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "# Dosage Extraction Function \nfrom pybgen import PyBGEN\ndef getVariantDosagesfromSnp(interestingSNPs):\n    sample_df = None\n    for snpOfChoice in interestingSNPs:\n        chrom, pos, ref, alt, snp = split_input(snpOfChoice)\n        imputed_path = getImputedPath(chrom)\n        wes_path = getWESPath(chrom)\n        chromNum = chrom[3:]\n        if chromNum == 'X':\n            chromNum = 23\n        with PyBGEN(imputed_path) as bgen: # Change to wes_path if using whole exome sequencing data\n            try:\n                variant = bgen.get_specific_variant(chrom, pos, ref, alt) #Use chromNum if WES, chrom if imputed\n            except ValueError:\n                print(f\"Variant {snp} not found. Skipping.\")\n                continue \n        if sample_df is None:\n            sample_df = getWESSampleDF(chrom)\n        chromosome = chrom \n        position = str(pos)\n        temp_df = pd.DataFrame(variant[0][1], columns= [snp])\n        dosage_df = pd.concat([sample_df, temp_df], axis=1)\n        sample_df = dosage_df\n    return dosage_df", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "from pybgen import PyBGEN\n# Converts a list of chr, pos, rsnum to a list of variants with alleles found in the bgen file \n# Specifically looks for all variants with dosage information available in given location including multiallelic variants, can ignore if chr:pos:ref:alt:rsnum is known \n\ndef getVariants(snplocs):\n    newsnps = []\n    foundsnps = []\n    for snp in snplocs:\n        chrom, pos, rsnum = snp.split(':')\n        imputed_path = getImputedPath(chrom)\n        wes_path = getWESPath(chrom)\n        chromNum = chrom[3:]\n        if chromNum == 'X':\n            chromNum = 23\n        with PyBGEN(imputed_path) as bgen: \n            val = bgen.iter_variants_in_region(chrom, pos, pos)\n            for v in val:\n                print(np.nansum(v[1]))\n                ref = (str(v[0]).split('_')[1].split('/')[0])\n                alt = (str(v[0]).split('_')[1].split('/')[1]).strip('>')\n                combined = ':'.join([chrom, pos, ref, alt, rsnum])\n                found = ':'.join([chrom, pos, rsnum])\n                #print(combined)\n                newsnps.append(combined)\n                foundsnps.append(found)\n    missedsnps = list(set(snplocs) - set(foundsnps))\n    for bad in missedsnps:\n        print(bad + ' not found')\n        \n    return newsnps\n            ", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "#ADRB1 Example\n\nADRB1_snps = ['chr10:114044277:A:G:rs1801252', \n             'chr10:114045297:G:C:rs1801253']\n\nmultipleSNPdosages = getVariantDosagesfromSnp(ADRB1_snps)\nmultipleSNPdosages", "metadata": {}, "execution_count": null, "outputs": []}]}